proposals/inline-classes.md

```diff
@@ -0,0 +1,163 @@
+
+* **Type**: Design proposal
+* **Author**: Mikhail Zarechenskiy
+* **Contributors**: Andrey Breslav, Denis Zharkov, Roman Elizarov, Stanislav Erokhin
+* **Status**: Under consideration
+* **Prototype**: Implemented in Kotlin 1.2.30
+
+Discussion of this proposal is held in [this issue](https://github.com/Kotlin/KEEP/issues/104).
+
+## Summary
+
+Currently, there is no performant way to create wrapper for a value of a corresponding type. The only way is to create a usual class, 
+but the use of such classes would require additional heap allocations, which can be critical for many use cases.    
+
+We propose to support identityless inline classes that would allow to introduce wrappers for a values without additional overhead related 
+to additional heap allocations.
+
+## Motivation / use cases
+
+Inline classes allow to create wrappers for a value of a certain type and such wrappers would be fully inlined. 
+This is similar to type aliases but inline classes are not assignment-compatible with the corresponding underlying types.
+
+Use cases:
+
+- Unsigned types
+```
+inline class UInt(private val value: Int) { ... }
+inline class UShort(private val value: Short) { ... }
+inline class UByte(private val value: Byte) { ... }
+inline class ULong(private val value: Long) { ... }
+```
+
+- Native types like `size_t` for Kotlin/Native
+- Inline enum classes 
+    - Int enum for [Android IntDef](https://developer.android.com/reference/android/support/annotation/IntDef.html)
+    - String enum for Kotlin/JS (see [WebIDL enums](https://www.w3.org/TR/WebIDL-1/#idl-enums))
+- Units of measurement
+- Result type (aka Try monad) [KT-18608](https://youtrack.jetbrains.com/issue/KT-18608)
+- Inline property delegates
+- Inline wrappers
+
+## Description
+
+Inline classes are declared using soft keyword `inline` and must have a single property:
+```
+inline class Foo(val i: Int)
+```
+Property `i` defines type of the underlying runtime representation for inline class `Foo`, while at compile time type will be `Foo`.
+
+From language point of view, inline classes can be considered as restricted classes, they can declare various members, operators, 
+have generics. 
+
+Example:
+```
+inline class Name(val s: String) : Comparable<Name> {
+    override fun compareTo(other: Name): Int = s.compareTo(other.s)
+    
+    fun greet() {
+        println("Hello, $s")
+    }
+}    
+
+fun greet() {
+    val name = Name("Kotlin") // there is no actual instantiation of class `Name`
+    name.greet() // method `greet` is called as a static method
+}
+```
+
+## Current limitations
+
+Currently, inline classes must satisfy the following requirements:
+
+- Inline class must have a public primary constructor with a single value parameter
+- Inline class must have a single read-only (`val`) property as an underlying value, which is defined in primary constructor
+- Underlying value cannot be of the same type that is containing inline class
+- Inline class cannot have `init` block
```

Comment:
Why would `init` block be prohibited? Isn't it where checks for the single value parameter would take place?

Reply:
You can't really check anything if at runtime the class isn't actually instantiated. Remember, it's `inline` after all.

Reply:
Because of Java interop. We can't ensure execution of `init` block when value comes from Java. Consider this case:
```
// file: sample.kt
inline class Positive(val u: Int) {
    init { assert(0 < u) }
}

fun foo(p: Positive) {} // in JVM bytecode `p` is just `int`

// Java
...
SampleKt.foo(-10) // it's OK from Java POV
```

Also, this is the reason for public primary constructor requirement.
I'll write about this in the proposal

Reply:
@zarechenskiy Why not rule out giving the underlying value of an inline class in Java at compile time? Java would only be allowed to receive inline classes values as the underlying type, but only Kotlin would be allowed to invoke the inline "constructor".

Reply:
I believe this question should be discussed in the [issue](https://github.com/Kotlin/KEEP/issues/104).
I see here several options: 
- we relax this requirement and have inconsistent behaviour
- we introduce some kind of assertions as for nullable values that come from Java
- use conversions for values that come from Java

Each of this option have cons, but it's worth to discuss it anyway